transcom/mymove
# B-19891 TAC LOA Invalid DB INT

## [Upstream branch](https://github.com/transcom/mymove/pull/13003)
- Also pulled in a commit per Cameron's recommendation from 19926: [adjust line of accounting handler to better handle no loas found](https://github.com/transcom/mymove/commit/8416fb8d6110818f49d419a57b309c143d4f5cad)
- 19926 is not an upstream

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-19891)

## Summary

This work adds in some logic checks that will determine the result of two flags that we want to store in the database. Advana gets an export of this table, so by storing these validity flags in the lines_of_accounting table, Advana will be able to see if any LOA information is missing.

Two columns that we store a validity check for on the `lines_of_accounting` table:
- `valid_hhg_code_for_loa`
- `valid_loa_for_tac`

Criteria for invalid LOA:

- No HHG Program Code
- No match to orders date and LOA eff date range (HHG)
- No match to orders date and current date (NTS)
- Missing any element per attachment in feature (DFAS elements)
- No LOA record found (LOA system ID key), or nothing returned based on existing logic (check logic for application of LOA to an EDI 858 for payment request transmission)

The logic checks that I added were specifically:
- No HHG Program Code
- Missing any element per attachment in feature (DFAS elements). See the screenshot or the attachment on the feature for the DFAS elements.
![image](https://github.com/transcom/mymove/assets/147537467/1c3bd9d4-1c62-46e3-8e53-7e47dc9199b8)

The other checks should be taken care of by the existing LOA and TAC lookup logic.

I also added a migration to move the valid_loa_for_tac column from the transportation_accounting_codes table to the lines_of_accounting table.

To Test:
1. Refresh your DB to pull in the migraiton.
2. Follow the instructions and use examples from #13003 to call the requestLineOfAccounting endpoint. Verify that the response contains:
`validHhgProgramCodeForLoa`
`validLoaForTac`
3. Verify that the `valid_hhg_code_for_loa` and `valid_loa_for_tac` columns in the database are populated with the correct values after calling the endpoint.
